sklearn/feature_selection/mutual_info_.py

```diff
@@ -0,0 +1,416 @@
+    if columns is None:
+        columns = range(X.shape[1])
+
+    for i in columns:
+        yield _get_column(X, i)
+
+
+def _estimate_mi(X, y, discrete_features='auto', discrete_target=False,
+                 n_neighbors=3, copy=True, random_state=None):
+    """Estimate mutual information between the features and the target.
+
+    Parameters
+    ----------
+    X : array_like or sparse matrix, shape (n_samples, n_features)
+        Feature matrix.
+    y : array_like, shape (n_samples,)
+        Target vector.
+    discrete_features : {'auto', bool, array_like}, default 'auto'
+        If bool, then determines whether to consider all features discrete
+        or continuous. If array, then it should be either a boolean mask
+        with shape (n_features,) or array with indices of discrete features.
+        If 'auto', it is assigned to False for dense `X` and to True for
+        sparse `X`.
+    discrete_target : bool, default False
+        Whether to consider `y` as a discrete variable.
+    n_neighbors : int, default 3
+        Number of neighbors to use for MI estimation for continuous variables,
+        see [1]_ and [2]_. Higher values reduce variance of the estimation, but
+        could introduce a bias.
+    copy : bool, default True
+        Whether to make a copy of the given data. If set to False, the initial
+        data will be overwritten.
+    random_state : int seed, RandomState instance or None, default None
+        The seed of the pseudo random number generator for adding small noise
+        to continuous variables in order to remove repeated values.
+
+    Returns
+    -------
+    mi : ndarray, shape (n_features,)
+        Estimated mutual information between each feature and the target.
+        A negative value will be replaced by 0.
+
+    References
+    ----------
+    .. [1] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    .. [2] B. C. Ross "Mutual Information between Discrete and Continuous
+           Data Sets". PLoS ONE 9(2), 2014.
+    """
+    X, y = check_X_y(X, y, accept_sparse='csc', y_numeric=not discrete_target)
+    n_samples, n_features = X.shape
+
+    if discrete_features == 'auto':
+        discrete_features = issparse(X)
+
+    if isinstance(discrete_features, bool):
+        discrete_mask = np.empty(n_features, dtype=bool)
+        discrete_mask.fill(discrete_features)
+    else:
+        discrete_features = np.asarray(discrete_features)
+        if discrete_features.dtype != 'bool':
+            discrete_mask = np.zeros(n_features, dtype=bool)
+            discrete_mask[discrete_features] = True
+        else:
+            discrete_mask = discrete_features
+
+    continuous_mask = ~discrete_mask
+    if np.any(continuous_mask) and issparse(X):
+        raise ValueError("Sparse matrix `X` can't have continuous features.")
+
+    if copy:
+        X = X.copy()
+
+    if not discrete_target and np.any(continuous_mask):
```

Comment:
can you remind me why we prefer not to scale X, if y is discrete? (and maybe add a comment)